cpp/dolfinx/io/ADIOS2_utils.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (C) 2024 Abdullah Mujahid, Jørgen S. Dokken and Garth N. Wells
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#ifdef HAS_ADIOS2
+
+#include <adios2.h>
+#include <cassert>
+#include <filesystem>
+#include <mpi.h>
+
+/// @file ADIOS2_utils.h
+/// @brief Utils for ADIOS2
+
+namespace dolfinx::io
+{
+
+/// ADIOS2-based writers/readers
+class ADIOS2Engine
+{
+public:
+  /// @brief Create an ADIOS2-based engine writer/reader
+  /// @param[in] comm The MPI communicator
+  /// @param[in] filename Name of output file
+  /// @param[in] tag The ADIOS2 object name
+  /// @param[in] engine ADIOS2 engine type. See
+  /// https://adios2.readthedocs.io/en/latest/engines/engines.html.
+  /// @param[in] mode ADIOS2 mode, default is Write or Read
+  ADIOS2Engine(MPI_Comm comm, const std::filesystem::path& filename,
+               std::string tag, std::string engine = "BP5",
+               const adios2::Mode mode = adios2::Mode::Write);
```

Comment:
Implicit write seems odd; no default?

Reply:
Default `mode` is `adios2::Mode::Write`

For a reader,  we need to specify `adios2::Mode::Read` explicitly.

Did you mean this or did I not understand your comment?